rejectOnEmpty mode

sushantdhiman opened the pull request:
Allow finder methods to throw `RecordNotFound` error when query returns null set and `kenophobic` mode is `true`. This option can be set per method basis or with model options.

Closes #5480 
Closes #272

> Kenophobic means a person with fear of emptiness or void, You can suggest any other name if want to :)

mickhansen commented:
While i like the name i think `throwOnEmpty` or `rejectOnEmpty` would be more fitting.
Could you please also extend the API to allow a custom error from the user as an argument? (Would be helpfull for people doing error handling in APIs wanting to pass a 404 erorr instance

mickhansen commented on lib/model.js:
```diff
@@ -1409,6 +1415,9 @@ Model.prototype.findAll = function(options) {
 Model.$findSeparate = function(results, options) {
+  if(_.isEmpty(results) && options.kenophobic)
+    throw new sequelizeErrors.RecordNotFoundError();
+
```
Lets place this in the find call. findSeperate deals with other stuff.

sushantdhiman commented:
Hi @mickhansen , I am handling the `kenophobic` mode from `findAll` now, not using the `$findSeparate` anymore. Also custom error instance can be passed to `kenophobic` flag and it will be thrown, added tests for that. Please review

mickhansen commented on lib/model.js:
```diff
@@ -1404,6 +1410,16 @@ Model.prototype.findAll = function(options) {
+    //kenophobic mode
+    if(_.isEmpty(results) && options.kenophobic) {
+      if( (typeof options.kenophobic === 'function')) {
+        throw new options.kenophobic();
```
It could also be an already init'e derror instance: `kenophobic: new Status404Error('No such user')`

sushantdhiman replied:
Not sure if I have done it in right way, but this will work for both cases now. Please check

mickhansen replied:
yeah looks good

mickhansen commented on lib/errors.js:
```diff
@@ -299,3 +299,15 @@ error.InstanceError = function (message) {
+  error.BaseError.apply(this, arguments);
+  this.name = 'SequelizeRecordNotFoundError';
+  this.message = message;
+};
```
`RecordNotFound` implies single result, maybe a better generic would be `EmptyResultError`

mickhansen commented:
We probably still need an option name change, /cc @janmeier

sushantdhiman commented:
I am good with of any of these options

* `rejectOnEmpty`
* `throwOnEmpty`
* `kenophobic`

`kenophobic` is bit inspired by `paranoid`, your comments @janmeier

mickhansen commented on lib/model.js:
```diff
@@ -1404,6 +1410,18 @@ Model.prototype.findAll = function(options) {
+
+    //kenophobic mode
+    if(_.isEmpty(results) && options.kenophobic) {
+      if( (typeof options.kenophobic === 'function')) {
```
space after `if` (goes for entire diff), no space after `(`, no need to have double `()` here actually.

mickhansen commented on test/integration/model/findAll.test.js:
```diff
@@ -1389,4 +1389,42 @@ describe(Support.getTestDialectTeaser('Model'), function() {
+          })).to.eventually.be.rejectedWith(Sequelize.ConnectionError);
+        });
+    });
+  });
```
We need an extra unit test so that we handle all 3 cases, none passed, class passed or instance passed

sushantdhiman commented:
@mickhansen I have fixed code styling issues, Added new tests to check cases of `Initialized`, `Instance` and `undefined` value for `kenophobic` flag

sushantdhiman commented:
@janmeier waiting for you :)

mickhansen commented:
Will give him another day or so and we'll just make a decision.

janmeier commented on lib/model.js:
```diff
@@ -1353,7 +1356,10 @@ Model.prototype.findAll = function(options) {
   options = optClone(options);
 
-  _.defaults(options, { hooks: true });
+  _.defaults(options, { hooks: true, kenophobic: false });
```
Couldn't this be ` _.defaults(options, { hooks: true, kenophobic: this.options.kenophobic });`?

sushantdhiman replied:
Yes it can be, it will save one line of code too

sushantdhiman replied:
Just tested, I didn't knew `defaults` behavior, but it will only overwrite `undefined` values, even if value is `false` it will not assign the `true` value from model options.

janmeier commented:
I lol'd at kenophobic, but I think `rejectOnEmpty` will be better ;)

sushantdhiman commented:
`rejectOnEmpty` :( , Ok I will do it

mickhansen commented:
@sushantdhiman Hey, why'd you close this? This was about ready to merge (just forgot about it since commits don't send me a notification :()

sushantdhiman commented:
Hi @mickhansen , I know you are busy but please review the pull requests when you get chance, if they are merged on time I feel encouraged to work on new one.

mickhansen commented:
Of course @sushantdhiman, but remember that commits don't notify me so i can't know you've made changes unless you comment.

If you can simply squash to a few descriptive commits (or change the descriptions to reflect the new name) we should be ready to go.

sushantdhiman commented:
Hi @mickhansen , I have changed the commit messages and squashed them

mickhansen commented:
Thanks @sushantdhiman, great work.

sushantdhiman commented:
Thanks @mickhansen
